### PR TITLE
Refactor Dockerfile sso

### LINKED ISSUE #1668: Apply Dockerfile best practices from Snyk recommendations
I saw a nice post today from Synk about best practices for securing node apps in Docker:

https://snyk.io/blog/10-best-practices-to-containerize-nodejs-web-applications-with-docker/

Screen Shot 2021-02-07 at 9 41 11 AM

We should apply as many of these as makes sense for us to our existing Dockerfiles, and also the new ones happening in #1636 for microservices.

### PR DESCRIPTION
**Pull request by Kevan-Y**
## Issue This PR Addresses
Fixes #1668

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [ ] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

- Add multi-staging layers
- Add Docker Healthcheck
- Use a smaller base image

Image is a bit bigger because we install the pnpm package
```bash
❯ docker images | grep refactor/sso
refactor/sso                                    new                d31f703b6c2b   18 seconds ago       132MB
refactor/sso                                    old                02baf2192da0   2 weeks ago          135MB
```

## Steps to test the PR

- `pnpm services:start`
- Go to `http://localhost/v1/auth/healthcheck`
## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: src/api/sso/Dockerfile [hunk 1]
```diff
@@ -1,47 +1,64 @@
...
 ENV NPM_CONFIG_COLOR=false
 
+RUN npm i -g pnpm
+
 # We'll run the app as the `node` user, so put it in their home directory
 WORKDIR /home/node/app
-# Copy the package.json and lock file over
-COPY package*.json /home/node/app/
+
+## Dependencies-dev ###############################################################
```
Comment (humphd) [inline src/api/sso/Dockerfile:14]: I don't think we're actually using this pattern anywhere else, so maybe we should remove this.

**Review by humphd — CHANGES_REQUESTED** [review]
Let's drop the dev/dependencies stuff, we don't need.

**Review by humphd — APPROVED** [review]
Looks good

**Comment by Kevan-Y** [comment]
Great, I'm gonna squash those commits.

**Review by AmasiaNalbandian — DISMISSED** [review]

**Review by humphd — CHANGES_REQUESTED** [review]
I can't build this on Mac (M1), I need python.

**Comment by Kevan-Y** [comment]
> I can't build this on Mac (M1), I need python.

Using a bigger node image for the building part, I think it should work for M1.

**Comment by humphd** [comment]
Let's get this rebased, there's been a lot of changes on the Dockerfiles recently.

**Comment by humphd** [comment]
I can't see what's failing on CI, and I can't reproduce the failure locally.  I'm stumped.

**Comment by Kevan-Y** [comment]
This is so weird, 
I checked out this branch and ran my test case on CentOS, but the test case is passing.
image

**Comment by humphd** [comment]
Really interesting.  I wonder what we're hitting?  Maybe it's as simple as resource issues, but I can't understand how either PR would affect it.  These are running on a VM with 2 vCPU and 7 GiB of memory.

I wonder if we're crashing the login container or something, but that wouldn't explain why subsequent tests on other browsers work.

We need to get more creative with our thinking...

**Comment by humphd** [comment]
This is what's running in that test: https://github.com/Seneca-CDOT/telescope/blob/a037690f2f55a7cb763e064b85dd4136ffec5e39/src/api/sso/test/e2e/browser-util.js#L14-L42

I wonder if we should add a bunch of logs so we can see where in the test it's getting stuck.

It could also be somehow related to https://github.com/Seneca-CDOT/telescope/blob/a037690f2f55a7cb763e064b85dd4136ffec5e39/src/api/sso/test/e2e/supabase-util.js#L8-L45

**Comment by Kevan-Y** [comment]
> @Kevan-Y @JiaHua-Zou can we summarize what we know about this failure here?

Failing in Github actions in `FAIL browser: chromium src/api/sso/test/e2e/auth-flows.test.js - Login flow preserves state param (25227 ms)`

Environment tested locally:
**Windows 11 pro 21H2** - test case pass.
**Windows 11 pro 21H2 on WSL2 Ubuntu-20.04** - test case pass
**Amazon Linux 2 AMI (HVM) - Kernel 5.10 (ami-0c02fb55956c7d316 (64-bit x86)) - t2.large** - test case pass 
**MacOS M1**, test case pass @humphd

**Comment by Kevan-Y** [comment]
> This is what's running in that test:
> 
> https://github.com/Seneca-CDOT/telescope/blob/a037690f2f55a7cb763e064b85dd4136ffec5e39/src/api/sso/test/e2e/browser-util.js#L14-L42
> 
> I wonder if we should add a bunch of logs so we can see where in the test it's getting stuck.
> 
> It could also be somehow related to
> 
> https://github.com/Seneca-CDOT/telescope/blob/a037690f2f55a7cb763e064b85dd4136ffec5e39/src/api/sso/test/e2e/supabase-util.js#L8-L45

Will try this and see what is the place that is stuck

**Comment by Kevan-Y** [comment]
Converting PR into a draft while I'm debugging the fail test case.

**Review by humphd — DISMISSED** [review]
Great work not giving up on this.

**Comment by Kevan-Y** [comment]
@humphd, sorry to make you re-review again, this is the right change with pnpm + adding Docker best practices

**Review by humphd — APPROVED** [review]

**Comment by humphd** [comment]
This i ready to go, just needs one more r+

**Review by aserputov — APPROVED** [review]

**Pull request merged**